### ISSUE TITLE
Scaling outputs

### ISSUE THREAD

**Issue by ahmeda14960** [ialhashim/DenseDepth#129]
Hello, I am using DenseDepth to predict depth from an image I generated from the simulation. The image is shown directly below:
![14](https://user-images.githubusercontent.com/30680697/83915206-cd294180-a740-11ea-94eb-92abf29195d9.png)

After looking through the linked paper, I found the following explanation of how to use the outputs from DenseDepth to predict real-world depth: 
![image](https://user-images.githubusercontent.com/30680697/83915723-cd760c80-a741-11ea-9b2f-6cd0532471b7.png)

Which means that  y_orig = m/y. I implemented this in the following code
```python
existing = 'kitti.h5'
custom_objects = {'BilinearUpSampling2D': BilinearUpSampling2D, 'depth_loss_function': None}
model = load_model(existing, custom_objects=custom_objects, compile=False)

# just test on one image
# convert this image to an RGB numpy array
image_files = ['14.png']
data = load_images(image_files)

# maximum depth for KITTI is 80 meters
m = 80
pred_arr = predict(model, data, 0, 80)
# output is (1, height, width, 1) for some reason
y = pred_arr[0, :, :, 0]
y_orig = m / y

plt.imshow(y_orig, cmap='magma_r')
plt.colorbar()
plt.show()
```

Using the predict function and other utility functions from this repo. I'll put the predict function below for reference
```python
def DepthNorm(x, maxDepth):
    return maxDepth / x

def predict(model, images, minDepth=10, maxDepth=1000, batch_size=2):
    # Support multiple RGBs, one RGB image, even grayscale 
    # I'm guessing this is the grayscale case
    if len(images.shape) < 3: images = np.stack((images,images,images), axis=2)
    # and this is for when we only have one image
    if len(images.shape) < 4: images = images.reshape((1, images.shape[0], images.shape[1], images.shape[2]))
    # Compute predictions
    predictions = model.predict(images, batch_size=batch_size)
    # Put in expected range
    return np.clip(DepthNorm(predictions, maxDepth=maxDepth), minDepth, maxDepth) / maxDepth
```

And I got the following output
![dense_depth](https://user-images.githubusercontent.com/30680697/83917699-53e01d80-a745-11ea-808b-cb790125bc9b.png)

This doesn't make sense to me for two reasons. 1) it seems to range from 0 to 1400 when it should range from 0 to 80, and 2) The objects in the foreground which are closer to the camera are given depth predictions that are much larger than objects in the background which are further from the camera. 

Am I using the functions from this library properly?

**Comment by TimbusCalin** [comment]
Your issue is that instead of multiplying * 80 at y = pred_arr[0, :, :, 0], y_true = y * 80, you apply the DepthNorm (reciprocal of Depth) twice.

You must multiply y = pred_arr[0, :, :, 0] with *80 in order to get the real depths.

PS: Did you make any change to the loss function when you trained? Mine looks like this (do not know if it's correct, but since my GT has real values between 1 and 80, maxDepthVal = 80.0:

```
def depth_loss_function(y_true, y_pred, theta=0.1, maxDepthVal=80.0):
    # Point-wise depth
    l_depth = K.mean(K.abs(y_pred - y_true), axis=-1)

    # Edges
    dy_true, dx_true = tf.image.image_gradients(y_true)
    dy_pred, dx_pred = tf.image.image_gradients(y_pred)
    l_edges = K.mean(K.abs(dy_pred - dy_true) + K.abs(dx_pred - dx_true), axis=-1)

    # Structural similarity (SSIM) index
    l_ssim = K.clip((1 - tf.image.ssim(y_true, y_pred, maxDepthVal)) * 0.5, 0, 1)

    # Weights
    w1 = 1.0
    w2 = 1.0
    w3 = theta

    return (w1 * l_ssim) + (w2 * K.mean(l_edges)) + (w3 * K.mean(l_depth))
```

**Comment by ahmeda14960** [comment]
Hello Timbus,

I don't think this is the case because with that change:
```python
existing = 'kitti.h5'
custom_objects = {'BilinearUpSampling2D': BilinearUpSampling2D, 'depth_loss_function': None}
model = load_model(existing, custom_objects=custom_objects, compile=False)

# just test on one image
# convert this image to an RGB numpy array
image_files = ['14.png']
data = load_images(image_files)


# maximum depth for KITTI is 80 meters
m = 80
pred_arr = predict(model, data, 0, 80)
# output is (1, height, width, 1) for some reason
y = pred_arr[0, :, :, 0]*m
y_orig = 1 / y

plt.imshow(y_orig, cmap='magma_r')
plt.colorbar()
plt.show()

mono_pred_arr = np.load('14_disp.npy')
ground_truth = np.load('14.npy')
ground_truth[ground_truth < 0] = 0
ground_truth[ground_truth > 80] = 80
pred_depth = 1 / mono_pred_arr[0, 0]
ratio = np.median(ground_truth) / np.median(pred_depth)
pred_depth *= ratio

plt.imshow(pred_depth, cmap='magma_r')
plt.colorbar()
plt.show()

plt.imshow(ground_truth, cmap='magma_r')
plt.colorbar()
plt.show()
```

I get the following.

![mul80](https://user-images.githubusercontent.com/30680697/84090467-e8ee4b00-a9bf-11ea-8f76-de774e1acea2.png)

Which is still not the correct range, and still shows the foreground as being further away than the background.

Also I see that the reciprocal of the depth is being applied twice here
```python
def DepthNorm(x, maxDepth):
    return maxDepth / x

def predict(model, images, minDepth=10, maxDepth=1000, batch_size=2):
    # Support multiple RGBs, one RGB image, even grayscale 
    # I'm guessing this is the grayscale case
    if len(images.shape) < 3: images = np.stack((images,images,images), axis=2)
    # and this is for when we only have one image
    if len(images.shape) < 4: images = images.reshape((1, images.shape[0], images.shape[1], images.shape[2]))
    # Compute predictions
    predictions = model.predict(images, batch_size=batch_size)
    # Put in expected range
    return np.clip(DepthNorm(predictions, maxDepth=maxDepth), minDepth, maxDepth) / maxDepth
```

But the above snippet isn't my code; it's a utility function from the densedepth library and it was used in an example file to evaluate a trained model on a test image. 

Also, I'm not sure I understand your last question because I didn't train a model, I used the pretrained KITTI model on this repository for inference on a single image.

**Comment by resindraburiza** [comment]
Hello,

I think you dont need to do anything to the output of the `predict` function. 
The prediction of the model is 
`predictions = model.predict(images, batch_size=batch_size)` which is in the form of reciprocal depth. Then the code returns 
`np.clip(DepthNorm(predictions, maxDepth=maxDepth), minDepth, maxDepth) / maxDepth`
which will bring the predicted value to the real scale.

**Comment by ahmeda14960** [comment]
Hello Aji,

Thank you for your suggestion! it seems to have fixed the problem where the relative depths seemed to have been inverted, but the scale is still off:
![pred_out](https://user-images.githubusercontent.com/30680697/84212194-28ce3480-aa8b-11ea-9da4-10d1a7553f01.png)
As it is not in the range [0, 80] as I expected. With that being said, the obvious thing for me to do here is min-max normalization which got me the following plot:
![scaled](https://user-images.githubusercontent.com/30680697/84213065-abf08a00-aa8d-11ea-8816-1892b42f8145.png)
This seems reasonable, but I'm left wondering if I'm overlooking/misunderstanding something in the code, because the predict function asked for a minimum and maximum value already, so I expected the model to output depths and not just disparities.

On a potentially unrelated note, I originally got an error that the image I was loading in had 4 channels when DenseDepth expects three channels, so I modified the "load_images" function from this repository by converting the image file into RGB format like so:
```python
def load_images(image_files):
    loaded_images = []
    for file in image_files:
        x = np.clip(np.asarray(Image.open( file ).convert("RGB"), dtype=float) / 255, 0, 1)
        loaded_images.append(x)
    return np.stack(loaded_images, axis=0)
```

Just thought I would include this in case this lead to some weird bug with scaling that I overlooked...

**Comment by resindraburiza** [comment]
Hello Ahmeda,

maybe the pretrained KITTI model was not trained with `minDepth= 0` and `MaxDepth=80` but something else? 
Also, correcting my initial answer, if we see the return of predict function, we see this:
`return np.clip(DepthNorm(predictions, maxDepth=maxDepth), minDepth, maxDepth) / maxDepth`

It means, the range of the predicted depth is [0,1] and if you want to get the real world scale you should multiply the predicted depth with the maxDepth.
```
# maximum depth for KITTI is 80 meters
pred_arr = predict(model, data, 0, 80)
pred_arr = [0,:,:,0]*80
```
But again, it's strange that the predicted depth is quite off. Do you have the ground truth depth of your simulation image? Maybe you can compare it. It is the best practice to check whether the estimated depth is correct or not.

About the edited load_images, you can always check the resulting x or image in runtime using a debugger.

Best regards,
Aji

**Comment by TimbusCalin** [comment]
@resindraburiza  This is what I was also saying; at the same time, If he has problems when multiplying *80, it's related to the Ground Truth Depth of the simulation image, like you said.

**Comment by resindraburiza** [comment]
Hi @TimbusCalin !
I only skimmed your comment before and I didn't realize that you already said what I wanted to say. Adding a proper credit to my answer :D

cheers!

**Comment by ahmeda14960** [comment]
@TimbusCalin  I didn't read your answer correctly, thank you for pointing it out again. After making the modification I get the following depth map
![timbus](https://user-images.githubusercontent.com/30680697/84427367-cc862480-abf2-11ea-922d-df06330e2646.png)

Which seems reasonable (up to some scaling factor). Thank you both for your help!

On the note of the simulation image, indeed there is something strange going on because this is the ground depth for the scene.
![ground_depth](https://user-images.githubusercontent.com/30680697/84427486-fdfef000-abf2-11ea-990b-ce200c4464f7.png)

I think what's going on is that I set the Field of View for my camera settings to be too small. It's currently 30 degrees, but a while back I was working with a FOV that was 90 degrees which is probably a better fit for this network. Regardless, the issue doesn't seem to be with DenseDepth but with how I collected my data.
